snapcraft/internal/pluginhandler/_runner.py

```diff
@@ -0,0 +1,193 @@
+    errors,
+)
+
+from ._scriptlet_functions import functions
+
+
+class Runner:
+    """The Runner class is responsible for orchestrating scriptlets."""
+
+    # FIXME: Need to quote builtin_functions typing because of
+    # https://github.com/python/typing/issues/259 which is fixed in Python
+    # 3.5.3.
+    def __init__(self, *, part_properties: Dict[str, Any], builddir: str,
+                 builtin_functions: 'Dict[str, Callable[..., None]]') -> None:
+        """Create a new Runner.
+        :param dict part_properties: YAML properties set for this part.
+        :param str builddir: The build directory for this part.
+        :param dict builtin_functions: Dict of builtin function names to
+                                       actual callables.
+        """
+        self._builddir = builddir
+        self._builtin_functions = builtin_functions
+
+        self._prepare_scriptlet = part_properties.get('prepare')
+        self._build_scriptlet = part_properties.get('build')
+        self._install_scriptlet = part_properties.get('install')
+
+    def prepare(self) -> None:
+        """Run prepare scriptlet."""
+        if self._prepare_scriptlet:
+            self._run_scriptlet(
+                'prepare', self._prepare_scriptlet, self._builddir)
+
+    def build(self) -> None:
+        """Run build scriptlet."""
+        if self._build_scriptlet:
+            self._run_scriptlet(
+                'build', self._build_scriptlet, self._builddir)
+
+    def install(self) -> None:
+        """Run install scriptlet."""
+        if self._install_scriptlet:
+            self._run_scriptlet(
+                'install', self._install_scriptlet, self._builddir)
+
+    def _run_scriptlet(self, scriptlet_name: str, scriptlet: str,
+                       workdir: str) -> None:
+        with tempfile.TemporaryDirectory() as tempdir:
+            call_fifo = _NonBlockingFifo(
+                path=os.path.join(tempdir, 'function_call'),
+                permissions=os.O_RDONLY)
+            feedback_fifo = _NonBlockingFifo(
+                path=os.path.join(tempdir, 'call_feedback'),
+                permissions=os.O_WRONLY)
+
+            script = textwrap.dedent("""\
+                {env}
+                {functions}
+                {scriptlet}
+            """.format(
+                env=common.assemble_env(), scriptlet=scriptlet,
+                functions=functions(
+                    call_fifo=call_fifo.path,
+                    feedback_fifo=feedback_fifo.path)))
+
+            process = subprocess.Popen(
+                ['/bin/sh', '-e', '-c', script], cwd=self._builddir)
+
+            status = None
+            try:
+                while status is None:
+                    function_call = call_fifo.read()
+                    if function_call:
+                        self._handle_builtin_function(
+                            scriptlet_name, function_call.strip())
+                        feedback_fifo.write('success\n')
```

Comment:
why do we write `success\n`?

Reply:
All that's actually required is a '\n' to get the [read](https://github.com/snapcore/snapcraft/pull/2002/files#diff-8b247ab4b649deb9ee488e09837eba39R33) to stop waiting for the call to be handled, but that felt a little too magical for my taste :stuck_out_tongue: . This could eventually grow into actually passing information back to the scriptlet from the builtins if we want.